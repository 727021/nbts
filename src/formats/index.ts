export * as nbt from './nbt'
export * as snbt from './snbt'
export * as json from './json'
