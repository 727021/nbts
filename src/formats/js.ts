import { Tag, TagType, quote } from '../util'

export const serialize = (tag: Tag, indentLevel = 0): string => {
    const { type, value } = tag
    const nameArg = tag.name === undefined ? '' : `, ${quote(tag.name)}`
    const indent = new Array(indentLevel).fill('\t').join('')

    let output = ''

    switch (type) {
        case TagType.BYTE:
            output = `${indent}Tag.byte(${value as number}${nameArg})`
            break
        case TagType.SHORT:
            output = `${indent}Tag.short(${value as number}${nameArg})`
            break
        case TagType.INT:
            output = `${indent}Tag.int(${value as number}${nameArg})`
            break
        case TagType.LONG:
            output = `${indent}Tag.long(${(value as bigint) + 'n'}${nameArg})`
            break
        case TagType.FLOAT:
            output = `${indent}Tag.float(${value as number}${nameArg})`
            break
        case TagType.DOUBLE:
            output = `${indent}Tag.double(${value as number}${nameArg})`
            break
        case TagType.BYTE_ARRAY:
            output = `${indent}Tag.byteArray(Buffer.from([${(
                value as Buffer
            ).join(',')}])${nameArg})`
            break
        case TagType.STRING:
            output = `${indent}Tag.string(${quote(value as string)}${nameArg})`
            break
        case TagType.LIST:
            output = `${indent}Tag.list([\n${(value as Tag[])
                .map(t => serialize(t, indentLevel + 1))
                .join(',\n')}\n${indent}]${nameArg})`
            break
        case TagType.COMPOUND:
            output = `${indent}Tag.compound([\n${(value as Tag[])
                .map(t => serialize(t, indentLevel + 1))
                .join(',\n')}\n${indent}]${nameArg})`
            break
        case TagType.INT_ARRAY:
            output = `${indent}Tag.intArray([${(value as number[]).join(
                ','
            )}]${nameArg})`
            break
        case TagType.LONG_ARRAY:
            output = `${indent}Tag.longArray([${(value as bigint[])
                .map(i => i.toString() + 'n')
                .join(',')}]${nameArg})`
            break
    }

    if (indentLevel === 0) {
        output =
            "import { Tag } from '@mcstatic/nbts'\n\nconst root = " +
            output +
            '\n'
    }

    return output
}
