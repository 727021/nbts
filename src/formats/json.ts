import { Tag, TagType, tagTypeName } from '../util'

const isLong = (value: string) => /^\d+L$/.test(value)

const jsonReplacer = (key: string, value: any) => {
    if (typeof value === 'bigint') return value.toString() + 'L'
    if (typeof value === 'object' && value.type === 'Buffer') return value.data
    if (key === 'type') return tagTypeName(value)
    return value
}

const jsonReviver = (key: string, value: any) => {
    if (typeof value === 'object' && value.type === TagType.BYTE_ARRAY) {
        value.value = Buffer.from(value.value)
    } else if (typeof value === 'string' && isLong(value)) {
        return BigInt(value.slice(0, value.length - 1))
    } else if (key === 'type') {
        return TagType[value] ?? value
    }
    return value
}

export const serialize = (root: Tag): string => {
    return JSON.stringify(root, jsonReplacer, 4)
}

export const deserialize = (raw: Buffer): Tag => {
    const str = raw.toString()
    return JSON.parse(str, jsonReviver)
}
