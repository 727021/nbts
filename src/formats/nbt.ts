import { Tag, TagType } from '../util'

export const serialize = (tag: Tag): Buffer => {
    switch (tag.type) {
        case TagType.BYTE: {
            const named = typeof tag.name === 'string'
            const buf = Buffer.allocUnsafe(1 + (named ? 2 + tag.name!.length : 0) + 1)
            let offset = 0
            buf.writeInt8(tag.type, offset)
            offset += 1
            if (named) {
                buf.writeUInt16BE(tag.name!.length, offset)
                offset += 2
                if (tag.name!.length > 0) {
                    Buffer.from(tag.name!).copy(buf, offset)
                    offset += tag.name!.length
                }
            }
            buf.writeInt8(tag.value as number, offset)
            return buf
        }
        case TagType.SHORT: {
            const named = typeof tag.name === 'string'
            const buf = Buffer.allocUnsafe(1 + (named ? 2 + tag.name!.length : 0) + 2)
            let offset = 0
            buf.writeInt8(tag.type, offset)
            offset += 1
            if (named) {
                buf.writeUInt16BE(tag.name!.length, offset)
                offset += 2
                if (tag.name!.length > 0) {
                    Buffer.from(tag.name!).copy(buf, offset)
                    offset += tag.name!.length
                }
            }
            buf.writeInt16BE(tag.value as number, offset)
            return buf
        }
        case TagType.INT: {
            const named = typeof tag.name === 'string'
            const buf = Buffer.allocUnsafe(1 + (named ? 2 + tag.name!.length : 0) + 4)
            let offset = 0
            buf.writeInt8(tag.type, offset)
            offset += 1
            if (named) {
                buf.writeUInt16BE(tag.name!.length, offset)
                offset += 2
                if (tag.name!.length > 0) {
                    Buffer.from(tag.name!).copy(buf, offset)
                    offset += tag.name!.length
                }
            }
            buf.writeInt32BE(tag.value as number, offset)
            return buf
        }
        case TagType.LONG: {
            const named = typeof tag.name === 'string'
            const buf = Buffer.allocUnsafe(1 + (named ? 2 + tag.name!.length : 0) + 8)
            let offset = 0
            buf.writeInt8(tag.type, offset)
            offset += 1
            if (named) {
                buf.writeUInt16BE(tag.name!.length, offset)
                offset += 2
                if (tag.name!.length > 0) {
                    Buffer.from(tag.name!).copy(buf, offset)
                    offset += tag.name!.length
                }
            }
            buf.writeBigInt64BE(tag.value as bigint, offset)
            return buf
        }
        case TagType.FLOAT: {
            const named = typeof tag.name === 'string'
            const buf = Buffer.allocUnsafe(1 + (named ? 2 + tag.name!.length : 0) + 4)
            let offset = 0
            buf.writeInt8(tag.type, offset)
            offset += 1
            if (named) {
                buf.writeUInt16BE(tag.name!.length, offset)
                offset += 2
                if (tag.name!.length > 0) {
                    Buffer.from(tag.name!).copy(buf, offset)
                    offset += tag.name!.length
                }
            }
            buf.writeFloatBE(tag.value as number, offset)
            return buf
        }
        case TagType.DOUBLE: {
            const named = typeof tag.name === 'string'
            const buf = Buffer.allocUnsafe(1 + (named ? 2 + tag.name!.length : 0) + 8)
            let offset = 0
            buf.writeInt8(tag.type, offset)
            offset += 1
            if (named) {
                buf.writeUInt16BE(tag.name!.length, offset)
                offset += 2
                if (tag.name!.length > 0) {
                    Buffer.from(tag.name!).copy(buf, offset)
                    offset += tag.name!.length
                }
            }
            buf.writeDoubleBE(tag.value as number, offset)
            return buf
        }
        case TagType.BYTE_ARRAY: {
            const named = typeof tag.name === 'string'
            const val = tag.value as Buffer
            const buf = Buffer.allocUnsafe(1 + (named ? 2 + tag.name!.length : 0) + 4 + val.length)
            let offset = 0
            buf.writeInt8(tag.type, offset)
            offset += 1
            if (named) {
                buf.writeUInt16BE(tag.name!.length, offset)
                offset += 2
                if (tag.name!.length > 0) {
                    Buffer.from(tag.name!).copy(buf, offset)
                    offset += tag.name!.length
                }
            }
            buf.writeInt32BE(val.length, offset)
            offset += 4
            val.copy(buf, offset)
            return buf
        }
        case TagType.STRING: {
            const named = typeof tag.name === 'string'
            const val = tag.value as string
            const buf = Buffer.allocUnsafe(1 + (named ? 2 + tag.name!.length : 0) + 2 + val.length)
            let offset = 0
            buf.writeInt8(tag.type, offset)
            offset += 1
            if (named) {
                buf.writeUInt16BE(tag.name!.length, offset)
                offset += 2
                if (tag.name!.length > 0) {
                    Buffer.from(tag.name!).copy(buf, offset)
                    offset += tag.name!.length
                }
            }
            buf.writeUInt16BE(val.length, offset)
            offset += 2
            Buffer.from(val, 'utf-8').copy(buf, offset)
            return buf
        }
        case TagType.LIST: {
            const named = typeof tag.name === 'string'
            const val = tag.value as Tag[]
            const tagType = val[0]?.type ?? TagType.END
            const buf = Buffer.allocUnsafe(1 + (named ? 2 + tag.name!.length : 0) + 1 + 4)
            let offset = 0
            buf.writeInt8(tag.type, offset)
            offset += 1
            if (named) {
                buf.writeUInt16BE(tag.name!.length, offset)
                offset += 2
                if (tag.name!.length > 0) {
                    Buffer.from(tag.name!).copy(buf, offset)
                    offset += tag.name!.length
                }
            }
            buf.writeInt8(tagType)
            offset += 1
            buf.writeInt32BE(val.length)
            const tagBuffers = val.map(serialize)
            return Buffer.concat([
                buf,
                ...tagBuffers
            ])
        }
        case TagType.COMPOUND: {
            const named = typeof tag.name === 'string'
            const val = tag.value as Tag[]
            const buf = Buffer.allocUnsafe(1 + (named ? 2 + tag.name!.length : 0))
            let offset = 0
            buf.writeInt8(tag.type, offset)
            offset += 1
            if (named) {
                buf.writeUInt16BE(tag.name!.length, offset)
                offset += 2
                if (tag.name!.length > 0) {
                    Buffer.from(tag.name!).copy(buf, offset)
                    offset += tag.name!.length
                }
            }
            const tagBuffers = val.map(serialize)
            return Buffer.concat([
                buf,
                ...tagBuffers,
                Buffer.from([TagType.END])
            ])
        }
        case TagType.INT_ARRAY: {
            const named = typeof tag.name === 'string'
            const val = tag.value as number[]
            const buf = Buffer.allocUnsafe(1 + (named ? 2 + tag.name!.length : 0) + 4 + val.length * 4)
            let offset = 0
            buf.writeInt8(tag.type)
            offset += 1
            if (named) {
                buf.writeUInt16BE(tag.name!.length)
                offset += 2
                if (tag.name!.length > 0) {
                    Buffer.from(tag.name!).copy(buf, offset)
                    offset += tag.name!.length
                }
            }
            buf.writeInt32BE(val.length, offset)
            offset += 4
            for (const i of val) {
                buf.writeInt32BE(i, offset)
                offset += 4
            }
            return buf
        }
        case TagType.LONG_ARRAY: {
            const named = typeof tag.name === 'string'
            const val = tag.value as bigint[]
            const buf = Buffer.allocUnsafe(1 + (named ? 2 + tag.name!.length : 0) + 4 + val.length * 8)
            let offset = 0
            buf.writeInt8(tag.type)
            offset += 1
            if (named) {
                buf.writeUInt16BE(tag.name!.length)
                offset += 2
                if (tag.name!.length > 0) {
                    Buffer.from(tag.name!).copy(buf, offset)
                    offset += tag.name!.length
                }
            }
            buf.writeInt32BE(val.length, offset)
            offset += 4
            for (const i of val) {
                buf.writeBigInt64BE(i, offset)
                offset += 8
            }
            return buf
        }
    }
    return Buffer.from([])
}

export const deserialize = (raw: Buffer): Tag => {
    throw new Error('Not implemented')
}
