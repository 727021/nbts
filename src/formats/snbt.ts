import { Tag, TagType, quote } from '../util'

export const serialize = (tag: Tag, indentLevel = 0): string => {
    const { type, value } = tag
    const name =
        tag.name !== undefined
            ? quote(tag.name)
            : indentLevel === 0
            ? "''"
            : 'None'
    const indent = new Array(indentLevel).fill('\t').join('')
    switch (type) {
        case TagType.BYTE:
            return `${indent}TAG_Byte(${name}): ${value as number}`
        case TagType.SHORT:
            return `${indent}TAG_Short(${name}): ${value as number}`
        case TagType.INT:
            return `${indent}TAG_Int(${name}): ${value as number}`
        case TagType.LONG:
            return `${indent}TAG_Long(${name}): ${(
                value as bigint
            ).toString()}L`
        case TagType.FLOAT:
            return `${indent}TAG_Float(${name}): ${value as number}`
        case TagType.DOUBLE:
            return `${indent}TAG_Double(${name}): ${value as number}`
        case TagType.BYTE_ARRAY:
            return `${indent}TAG_Byte_Array(${name}): [${(value as Buffer).join(
                ', '
            )}]`
        case TagType.STRING:
            return `${indent}TAG_String(${name}): ${quote(value as string)}`
        case TagType.LIST: {
            const val = value as Tag[]
            const entries = `${val.length} ${
                val.length === 1 ? 'entry' : 'entries'
            }`
            let str = `${indent}TAG_List(${name}): ${entries}`
            if (val.length > 0) {
                str +=
                    `\n${indent}{` +
                    `\n${val
                        .map(t => serialize(t, indentLevel + 1))
                        .join('\n')}` +
                    `\n${indent}}`
            } else {
                str += ' {}'
            }
            return str
        }
        case TagType.COMPOUND: {
            const val = value as Tag[]
            const entries = `${val.length} ${
                val.length === 1 ? 'entry' : 'entries'
            }`
            let str = `${indent}TAG_Compound(${name}): ${entries}`
            if (val.length > 0) {
                str +=
                    `\n${indent}{` +
                    `\n${val
                        .map(t => serialize(t, indentLevel + 1))
                        .join('\n')}` +
                    `\n${indent}}`
            } else {
                str += ' {}'
            }
            return str
        }
        case TagType.INT_ARRAY:
            return `${indent}TAG_Int_Array(${name}): [${(
                value as number[]
            ).join(', ')}]`
        case TagType.LONG_ARRAY:
            return `${indent}TAG_Long_Array(${name}): [${(value as bigint[])
                .map(v => v.toString() + 'L')
                .join(', ')}]`
    }
    return ''
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const deserialize = (raw: Buffer): Tag => {
    throw new Error('Not implemented')
}
