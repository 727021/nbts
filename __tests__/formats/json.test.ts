import { deserialize, serialize } from '../../src/formats/json'
import { bigTest, helloWorld, playerNanValue } from '../data/test-data'

describe('json serializer', () => {
    it('serializes hello world correctly', () => {
        const result = serialize(helloWorld)
        expect(result).toMatchSnapshot()
    })

    it('serializes bigtest correctly', () => {
        const result = serialize(bigTest)
        expect(result).toMatchSnapshot()
    })

    it('serializes player-nan-value correvtly', () => {
        const result = serialize(playerNanValue)
        expect(result).toMatchSnapshot()
    })
})

describe('json deserializer', () => {
    it('deserializes hello world correctly', () => {
        const json = serialize(helloWorld)
        const result = deserialize(Buffer.from(json))
        expect(result).toEqual(helloWorld)
    })

    it('deserializes bigtest correctly', () => {
        const json = serialize(bigTest)
        const result = deserialize(Buffer.from(json))
        expect(result).toEqual(bigTest)
    })

    it('deserializes player-nan-value correctly', () => {
        const json = serialize(playerNanValue)
        const result = deserialize(Buffer.from(json))
        expect(result).toEqual(playerNanValue)
    })
})
