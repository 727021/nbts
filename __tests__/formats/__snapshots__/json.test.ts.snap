// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json serializer serializes bigtest correctly 1`] = `
"{
    "type": "COMPOUND",
    "value": [
        {
            "type": "LONG",
            "value": "9223372036854775807L",
            "name": "longTest"
        },
        {
            "type": "SHORT",
            "value": 32767,
            "name": "shortTest"
        },
        {
            "type": "STRING",
            "value": "HELLO WORLD THIS IS A TEST STRING ÅÄÖ!",
            "name": "stringTest"
        },
        {
            "type": "FLOAT",
            "value": 0.4982314705848694,
            "name": "floatTest"
        },
        {
            "type": "INT",
            "value": 2147483647,
            "name": "intTest"
        },
        {
            "type": "COMPOUND",
            "value": [
                {
                    "type": "COMPOUND",
                    "value": [
                        {
                            "type": "STRING",
                            "value": "Hampus",
                            "name": "name"
                        },
                        {
                            "type": "FLOAT",
                            "value": 0.75,
                            "name": "value"
                        }
                    ],
                    "name": "ham"
                },
                {
                    "type": "COMPOUND",
                    "value": [
                        {
                            "type": "STRING",
                            "value": "Eggbert",
                            "name": "name"
                        },
                        {
                            "type": "FLOAT",
                            "value": 0.5,
                            "name": "value"
                        }
                    ],
                    "name": "egg"
                }
            ],
            "name": "nested compound test"
        },
        {
            "type": "LIST",
            "value": [
                {
                    "type": "LONG",
                    "value": "11L"
                },
                {
                    "type": "LONG",
                    "value": "12L"
                },
                {
                    "type": "LONG",
                    "value": "13L"
                },
                {
                    "type": "LONG",
                    "value": "14L"
                },
                {
                    "type": "LONG",
                    "value": "15L"
                }
            ],
            "name": "listTest (long)"
        },
        {
            "type": "LIST",
            "value": [
                {
                    "type": "COMPOUND",
                    "value": [
                        {
                            "type": "STRING",
                            "value": "Compound tag #0",
                            "name": "name"
                        },
                        {
                            "type": "LONG",
                            "value": "1264099775885L",
                            "name": "created-on"
                        }
                    ]
                },
                {
                    "type": "COMPOUND",
                    "value": [
                        {
                            "type": "STRING",
                            "value": "Compound tag #1",
                            "name": "name"
                        },
                        {
                            "type": "LONG",
                            "value": "1264099775885L",
                            "name": "created-on"
                        }
                    ]
                }
            ],
            "name": "listTest (compound)"
        },
        {
            "type": "BYTE",
            "value": 127,
            "name": "byteTest"
        },
        {
            "type": "BYTE_ARRAY",
            "value": [
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48
            ],
            "name": "byteArrayTest (the first 1000 values of (n*n*255+n*7)%100, starting with n=0 (0, 62, 34, 16, 8, ...))"
        },
        {
            "type": "DOUBLE",
            "value": 0.4931287132182315,
            "name": "doubleTest"
        }
    ],
    "name": "Level"
}"
`;

exports[`json serializer serializes hello world correctly 1`] = `
"{
    "type": "COMPOUND",
    "value": [
        {
            "type": "STRING",
            "value": "Bananrama",
            "name": "name"
        }
    ],
    "name": "hello world"
}"
`;

exports[`json serializer serializes player-nan-value correvtly 1`] = `
"{
    "type": "COMPOUND",
    "value": [
        {
            "type": "LIST",
            "value": [
                {
                    "type": "DOUBLE",
                    "value": 0
                },
                {
                    "type": "DOUBLE",
                    "value": 0
                },
                {
                    "type": "DOUBLE",
                    "value": 0
                }
            ],
            "name": "Motion"
        },
        {
            "type": "FLOAT",
            "value": 0,
            "name": "FallDistance"
        },
        {
            "type": "LIST",
            "value": [
                {
                    "type": "DOUBLE",
                    "value": 0
                },
                {
                    "type": "DOUBLE",
                    "value": 0
                },
                {
                    "type": "DOUBLE",
                    "value": 0
                }
            ],
            "name": "Pos"
        },
        {
            "type": "SHORT",
            "value": 20,
            "name": "Health"
        },
        {
            "type": "SHORT",
            "value": 0,
            "name": "DeathTime"
        },
        {
            "type": "SHORT",
            "value": -20,
            "name": "Fire"
        },
        {
            "type": "SHORT",
            "value": 300,
            "name": "Air"
        },
        {
            "type": "BYTE",
            "value": 1,
            "name": "OnGround"
        },
        {
            "type": "SHORT",
            "value": 0,
            "name": "HurtTime"
        },
        {
            "type": "LIST",
            "value": [
                {
                    "type": "FLOAT",
                    "value": 164.3999481201172
                },
                {
                    "type": "FLOAT",
                    "value": -63.150203704833984
                }
            ],
            "name": "Rotation"
        },
        {
            "type": "SHORT",
            "value": 0,
            "name": "AttackTime"
        },
        {
            "type": "LIST",
            "value": [],
            "name": "Inventory"
        }
    ],
    "name": ""
}"
`;
